Issue #70 - Add Serverless Support

currantw opened the pull request:
### Description

Adds support for OpenSearch Serverless to the ODBC driver.
 
### Issues Resolved

#70 
 
### Check List

- [x] New functionality includes testing.
_Automated testing has been deferred to future issue (see below). See comments for summary of manual testing._
- [x] All tests pass, including unit test, integration test and doctest.
_Pipeline job `build-mac` is failing. This is an existing failure on `main`, and I have raised a new issue to fix it (see below)_
- [x] New functionality has been documented.
- [x] New functionality has javadoc added - _Not applicable_ 
- [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### Related Issues

- #74 
- #75
- #76

currantw commented on src/sqlodbc/dlg_specific.c:
```diff
@@ -37,20 +37,43 @@ void makeConnectString(char *connect_string, const ConnInfo *ci, UWORD len) {
     /* fundamental info */
     nlen = MAX_CONNECT_STRING;
+
+    const char* connect_format_string =
+        "%s=%s;"
+        INI_SERVER "=%s;"
+        "database=OpenSearch;"
+        INI_PORT "=%s;"
+        INI_USERNAME_ABBR "=%s;"
+        INI_PASSWORD_ABBR "=%s;"
+        INI_AUTH_MODE "=%s;"
+        INI_REGION "=%s;"
+        INI_TUNNEL_HOST "=%s;"
+        INI_IS_SERVERLESS "=%d"
+        INI_SSL_USE "=%d;"
+        INI_SSL_HOST_VERIFY "=%d;"
+        INI_LOG_LEVEL "=%d;"
+        INI_LOG_OUTPUT "=%s;"
+        INI_TIMEOUT "=%s;"
+        INI_FETCH_SIZE "=%s;";
```
Just pulled this out to try to make it a bit more readable.

jduo commented on src/sqlodbc/opensearch_communication.cpp:
```diff
@@ -1048,15 +1071,38 @@ std::string OpenSearchCommunication::GetClusterName() {
+bool OpenSearchCommunication::isServerless() {
+
+    // Specified in DSN configuration.
+    const std::string& is_serverless = m_rt_opts.conn.is_serverless;
```
Can we make the is_serverless field itself a boolean and evaluate it earlier? Rather than every time isServerless() is called? It looks like isServerless() is called during IssueRequest() which is used repeatedly (for executing queries and paging through results).

currantw replied:
Good call. Added `is_serverless` member variable and a corresponding `SetIsServerless` method, which is called during `EstablishConnection` - so only once.

jduo commented on src/sqlodbc/opensearch_communication.h:
```diff
@@ -81,6 +81,7 @@ class OpenSearchCommunication {
     void SetErrorDetails(std::string reason, std::string message,
                          ConnErrorType error_type);
     void SetErrorDetails(ErrorDetails details);
+    bool isServerless();
```
Probably should be marked const since it shouldn't mutate state. Capital 'i' as well.

currantw replied:
Overtaken by subsequent changes. Has been replaced with non-`const` `SetIsServerless` method.

currantw commented:
Note: The `build_mac` action is also failing on the main project (https://github.com/opensearch-project/sql-odbc). Currently, building locally on Mac fails unless the `-Werror` flag is removed, so this is perhaps the cause of this failure. I will be raising an issue to fix this.

jduo commented on src/sqlodbc/opensearch_communication.cpp:
```diff
@@ -548,26 +564,31 @@ bool OpenSearchCommunication::CheckSQLPluginAvailability() {
+    SetIsServerless();
+    SetSqlEndpoint();
+
+    if (sql_endpoint == SQL_ENDPOINT_ERROR) {
```
Can we add a comment here clarifying how connecting to serverless affects these checks? It's hard to follow.

currantw replied:
Added some comments. Also update this `if` clause to also check `is_serverless` (even though this is unnecessary, it hopefully makes the dependencies a bit more clear). Let me know what you think!

currantw commented on docs/user/configuration_options.md:
```diff
@@ -27,9 +27,10 @@
+| `HostnameVerification` | Whether certificate hostname verification should be performed for an SSL/TLS connection.                                                 | boolean | `0` or `1`                      | `1` (true)  |
 | `ResponseTimeout`      | The maximum time to wait for responses from the `Host`, in seconds.                                                                      | integer |                                 | `10`        |
 | `FetchSize`            | The page size for all cursor requests. The default value (-1) uses server-defined page size. Set FetchSize to 0 for non-cursor behavior. | integer | `-1`, `0` or any positive value | `-1`        |
+| `IsServerless`         | Whether the connection is to an OpenSearch Serverless cluster. If not specified, it is determined by parsing the server URL.             | boolean | `0` or `1`                      |             | 
```
Renamed to `ServerlessOverride`.

currantw commented:
Automated testing is non-trivial, and has been deferred to #75. In the meantime, we have performed the following to test this new behaviour on Mac:

1. Created a new OpenSearch Serverless cluster that includes SQL.
2. Updated my AWS credentials file ( to add `opensearchodbc` profile with the appropriate access key, secret access key, and session token.
3. Load some test data to OpenSearch using Postman.
4. Downloaded the repo and build the `.dll` using CMake.
5. Updated my `odbc.ini` and `odbcinst.ini` files to configure the ODBC (see below)
6. Download [iodbctest](https://linux.die.net/man/1/iodbctest) (`brew install libiodbc`).
7. Run `iodbctest`, connect to the corresponding DSN, and query the test data.

<img width="1083" alt="image" src="https://github.com/user-attachments/assets/66cf0305-e6c5-48fd-9071-810a08835607">

```
# odbc.ini

[ODBC Data Sources]
OpenSearchServerlessDSN = OpenSearchODBC

[OpenSearchServerlessDSN]
Host                 = https://<HOST-URL>.aoss.amazonaws.com
Port                 = 443
UseSSL               = 1
ResponseTimeout      = 10
HostnameVerification = 1
Region               = <AWS_REGION>
Auth                 = AWS_SIGV4
LogOutput            = /Library/ODBC/opensearch-sql-odbc
Driver               = <PROJECT_ROOT>/build/odbc/lib/libsqlodbc.dylib
ServerlessOverride   = 1

# odbcinst.ini
[ODBC Drivers]
OpenSearchODBC = Installed

[OpenSearchODBC]
Driver = <PROJECT_ROOT>/build/odbc/lib/libsqlodbc.dylib
Setup  = <PROJECT_ROOT>//build/odbc/lib/libsqlodbc.dylib
```

acarbonetto commented on src/sqlodbc/opensearch_communication.cpp:
```diff
@@ -1046,17 +1075,41 @@ std::string OpenSearchCommunication::GetClusterName() {
-    } else if (distribution.compare("opensearch") == 0) {
-        sql_endpoint = "/_plugins/_sql";
+        sql_endpoint = SQL_ENDPOINT_ERROR;
+    } else if (distribution == "opensearch") {
```
defined "opensearch" as a static constant on the top of the file

currantw replied:
Thanks. Done ✅ 

Aside: Would probably be more robust to also add a `DISTRIBUTION_ELASTICSEARCH` constant as well, and then restructure to below. But I don't know what the constant would be for ElasticSearch, and don't want to get side-tracked by this.

```
std::string distribution = GetServerDistribution();
  if (distribution == DISTRIBUTION_OPENSEARCH) {
    sql_endpoint = SQL_ENDPOINT_OPENSEARCH;
  } else if (distribution == DISTRIBUTION_ELASTICSEARCH) {
    sql_endpoint = SQL_ENDPOINT_ELASTICSEARCH;
  } else {
    sql_endpoint = SQL_ENDPOINT_ERROR;
  }
```

acarbonetto commented on src/sqlodbc/opensearch_communication.h:
```diff
@@ -82,13 +81,17 @@ class OpenSearchCommunication {
     ConnStatusType m_status;
     ConnErrorType m_error_type;
     std::shared_ptr< ErrorDetails > m_error_details;
+    bool is_serverless;
```
why not is_aoss_serverless?

currantw replied:
I don't feel strongly about it. Serverless only exists as part of Amazon OpenSearch Service, so the "aoss" doesn't necessary add anything, but not a bad idea to be more explicit/tie it more clearly to the `aoss.amazonaws.com` URL component. 🤷

acarbonetto replied:
Always be explicit where possible.

currantw replied:
Updated to `is_aoss_serverless`. Also updated corresponding variable, constant, and function names for consistency.

currantw commented on src/sqlodbc/opensearch_types.h:
```diff
@@ -272,6 +272,7 @@ typedef struct connection_options {
     std::string port;
     std::string timeout;
     std::string fetch_size;
+    std::string is_serverless;
```
Done! ✅

currantw commented:
> Automated testing is non-trivial, and has been deferred to #75. In the meantime, we have performed the following to test this new behaviour on Mac:
> 
> **TESTING SERVERLESS**
> 
> Amazon OpenSearch Serverless cluster was setup for testing. See comments for more details.
> 
> ✅ ODBC connects successfully with `iodbctest`.
> 
> **TESTING NON-SERVERLESS WITH EDITED HOST FILE**
> 
> A local OpenSearch cluster was setup for testing.
> 
> ✅ ODBC connects successfully with `iodbctest`.
> 
> **NOTES ON TESTING**
> 
> The following steps where taken to setup and test the Serverless cluster:
> 
> 1. Created a new OpenSearch Serverless cluster that includes SQL.
> 2. Updated my AWS credentials file ( to add `opensearchodbc` profile with the appropriate access key, secret access key, and session token.
> 3. Load some test data to OpenSearch using Postman.
> 4. Downloaded the repo and build the `.dll` using CMake.
> 5. Updated my `odbc.ini` and `odbcinst.ini` files to configure the ODBC (see below)
> 6. Download [iodbctest](https://linux.die.net/man/1/iodbctest) (`brew install libiodbc`).
> 7. Run `iodbctest`, connect to the corresponding DSN, and query the test data.
> 
> <img alt="image" width="1083" src="https://private-user-images.githubusercontent.com/181785993/387397906-66cf0305-e6c5-48fd-9071-810a08835607.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxNjEwMTIsIm5iZiI6MTczMjE2MDcxMiwicGF0aCI6Ii8xODE3ODU5OTMvMzg3Mzk3OTA2LTY2Y2YwMzA1LWU2YzUtNDhmZC05MDcxLTgxMGEwODgzNTYwNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyMVQwMzQ1MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZGFiNGRiM2RmYWY5ZGMxM2IxMTE3NDM4ZGZmODY5M2FkZTRlYmRhYzYxZmRmNWFkZDAxMzhkY2E4MTgzMDliJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pAk68QmwD1HGsjHNI3Lj82fFt5EoM_JogbmL36H6a1g">
> ```
> # odbc.ini
> 
> [ODBC Data Sources]
> OpenSearchServerlessDSN = OpenSearchODBC
> 
> [OpenSearchServerlessDSN]
> Host                 = https://<HOST-URL>.aoss.amazonaws.com
> Port                 = 443
> UseSSL               = 1
> ResponseTimeout      = 10
> HostnameVerification = 1
> Region               = <AWS_REGION>
> Auth                 = AWS_SIGV4
> LogOutput            = /Library/ODBC/opensearch-sql-odbc
> Driver               = <PROJECT_ROOT>/build/odbc/lib/libsqlodbc.dylib
> ServerlessOverride   = 1
> 
> # odbcinst.ini
> [ODBC Drivers]
> OpenSearchODBC = Installed
> 
> [OpenSearchODBC]
> Driver = <PROJECT_ROOT>/build/odbc/lib/libsqlodbc.dylib
> Setup  = <PROJECT_ROOT>//build/odbc/lib/libsqlodbc.dylib
> ```
